### PR TITLE
feat: Build empty cookiecutters and run lint task during CI

### PR DESCRIPTION
**Pull request by flexponsive**
Related to #993

# Feature Scope

Adds a Github Action to run basic end-to-end tests for the cookiecutters.

It builds empty cookiecutters from [replay files](https://cookiecutter.readthedocs.io/en/stable/advanced/replay.html), install its dependencies with poetry and runs some auto fixes on the generated code (flake8 & co). Finally, it runs the lint task *as specified by the cookiecutter* against itself. Additionally, a script which creates and lints in one step for developer convenience while working on cookiecutter templates.

This PR also includes a small shell script to build, install and lint a cookiecutter with a single command. 

This PR also includes a few fixes (method signatures #1224, docstrings etc.) demanded by the linter for the case of a tap with REST stream type and API Key authentication method.

# Relationship to `test_cookiecutter.py`

The SDK test suite already includes tests for building cookiecutters in `test_cookiecutter.py`, which runs flake8 and mypy. However, it does not use the same configuration for these tools as the template contains and also does not install dependencies with poetry so some errors are not picked up (e.g. method signatures mismatch). End-to-end tests can be a useful addition here.

# Suggested GIthub Action config

Run the `cookiecutter-e2e` action only when cookiecutter files are changed; optionally on a regular schedule.

----
:books: Documentation preview :books:: https://meltano-sdk--1410.org.readthedocs.build/en/1410/

### CODE REVIEW

**Comment by tayloramurphy** [comment]
Thanks for the PR @flexponsive - looks like we've got some engineers assigned as reviewers who will get back to you likely sometime next week!

**Comment by edgarrmondragon** [comment]
Hey @flexponsive! Looks like flake8 is complaining. Let me know when this is ready to be reviewed 🙂

**Comment by flexponsive** [comment]
@edgarrmondragon I have updated the PR and description. Github does not show the results from the E2E-Tests of cookiecutters here in the PR, they are available here:
- [tap, rest, API key](https://github.com/flexponsive/meltano-sdk/actions/runs/4156469784/jobs/7190280635)
- [target, per record](https://github.com/flexponsive/meltano-sdk/actions/runs/4156469784/jobs/7190280688) - does not actually lint due to #360

**Comment by aaronsteers** [comment]
@flexponsive - First of all - _thank you_ for creating this PR! 

What do you think of adding a step to upload the generated projects using GitHub artifacts, if not already doing so?

https://github.com/actions/upload-artifact

```yml
- uses: actions/upload-artifact@v3
  with:
    name: my-artifact
    path: path/to/artifact/world.txt
```

File: .github/workflows/cookiecutter-e2e.yml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    - name: Check out the repository
+      uses: actions/checkout@v3.3.0
+
+    - name: Setup Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v4.5.0
+      with:
+        python-version: ${{ matrix.python-version }}
+        architecture: x64
+        cache: 'pip'
+        cache-dependency-path: 'poetry.lock'
```
Comment (edgarrmondragon) [inline .github/workflows/cookiecutter-e2e.yml:36]: Probably doesn't make sense to use the package's `poetry.lock` as cache key for the templates, e.g. the templates cache could need to refreshed even if `poetry.lock` hasn`t changed.
```suggestion
```

File: .github/workflows/cookiecutter-e2e.yml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    - name: Upgrade pip
+      env:
+        PIP_CONSTRAINT: .github/workflows/constraints.txt
+      run: |
+        pip install pip
+        pip --version
+
```
Comment (edgarrmondragon) [inline .github/workflows/cookiecutter-e2e.yml:44]: Extra line break

```suggestion
```

File: e2e-tests/cookiecutters/test_cookiecutter.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+CC_TEST_OUTPUT=$CC_BUILD_PATH/$CC_OUTPUT_DIR
+if [[ -d "$CC_TEST_OUTPUT" ]]; then
+    rm -fr "$CC_TEST_OUTPUT"
+fi
+#echo cookiecutter --replay-file $REPLAY_FILE $TAP_TEMPLATE -o $CC_BUILD_PATH; exit
+cookiecutter --replay-file $REPLAY_FILE $TAP_TEMPLATE -o $CC_BUILD_PATH &&
+    cd $CC_TEST_OUTPUT &&
+    pwd &&
+    poetry install
```
Comment (edgarrmondragon) [inline e2e-tests/cookiecutters/test_cookiecutter.sh:35]: Is this maybe missing a `poetry lock` step?

**Review by edgarrmondragon — CHANGES_REQUESTED** [review]
Hey @flexponsive. A few comments and suggestions. The bash script is good for an initial iteration of this but we should probably (not necessarily) in this PR port it to Nox since that's where we centralize dev tasks.

**Comment by flexponsive** [comment]
Thank you for the fast feedback and improvements to the PR!

@edgarrmondragon I accepted the suggestions and believe the issues you raised are fixed. I agree that it would be a good idea in the future to migrate from the bash script to tox but probably outside the scope of this PR.

@aaronsteers build artifact upload has been added too

Sample run: https://github.com/flexponsive/meltano-sdk/actions/runs/4172051621

**Comment by aaronsteers** [comment]
> build artifact upload has been added too

Nice! Thank you!

**Comment by edgarrmondragon** [comment]
Reviewing now...

**Review by edgarrmondragon — APPROVED** [review]

**Review by edgarrmondragon — APPROVED** [review]

**Comment by edgarrmondragon** [comment]
Thanks @flexponsive!

**Pull request merged**